demos/https/iot_demo_https_s3_download_async.c

```diff
@@ -0,0 +1,804 @@
+ * other asynchronous requests on the same connection. You can reuse the buffer only after the request/response process
+ * has been finished. It is finished when the responseCompleteCallback() is invoked. We create a pool of memory so that
+ * all available requests in the pool can be scheduled right away without over-writing each other. */
+#ifndef IOT_HTTPS_DEMO_MAX_ASYNC_REQUESTS
+    #define IOT_HTTPS_DEMO_MAX_ASYNC_REQUESTS    ( 3 )
+#endif
+
+/* Size in bytes of the User Buffer used to store the internal connection context. The size presented here accounts for
+ * storage of the internal connection context. The minimum size can be found in extern const unint32_t
+ * connectionUserBufferMinimumSize. */
+#ifndef IOT_DEMO_HTTPS_CONN_BUFFER_SIZE
+    #define IOT_DEMO_HTTPS_CONN_BUFFER_SIZE    ( 400 )
+#endif
+
+/* Size in bytes of the user buffer used to store the internal request context and HTTP request header lines.
+ * The size presented here accounts for the storage of the internal context, the first request line in the HTTP
+ * formatted header and extra headers. The minimum size can be found in extern const uint32_t
+ * requestUserBufferMinimumSize. Keep in mind that this requestUserBufferMinimumSize does not include the size of the
+ * path in the request line. The path could be well over 100 characters long as it includes not only the object key name
+ * in S3, but also the query following. The query following has the AWSAccessKeyId, the expiration time, and the
+ * AWS Signature Version 4 signature.*/
+#ifndef IOT_DEMO_HTTPS_REQ_USER_BUFFER_SIZE
+    #define IOT_DEMO_HTTPS_REQ_USER_BUFFER_SIZE    ( 512 )
+#endif
+
+/* Size in bytes of the user buffer used to store the internal response context and the HTTP response header lines.
+ * The size presented here accounts for the storage of the internal context, the first request line in the HTTP
+ * formatted header and extra headers. The minimum can be found in requestUserBufferMinimumSize.
+ * Keep in mind that if the headers from the response do not all fit into this buffer, then the rest of the headers
+ * will be discarded. The minimum size can be found in extern const uint32_t responseUserBufferMinimumSize. */
+#ifndef IOT_DEMO_HTTPS_RESP_USER_BUFFER_SIZE
+    #define IOT_DEMO_HTTPS_RESP_USER_BUFFER_SIZE    ( 1024 )
+#endif
+
+/* Size in bytes of the buffer used to store the response body (parts of it). This should be greater than or equal to
+ * the size of the file we want to download.*/
+#ifndef IOT_DEMO_HTTPS_RESP_BODY_BUFFER_SIZE
+    #define IOT_DEMO_HTTPS_RESP_BODY_BUFFER_SIZE    ( 512 )
+#endif
+
+/* Timeout in milliseconds to wait for all asynchronous requests to finish. This timeout starts when the last
+ * IOT_HTTPS_DEMO_MAX_ASYNC_REQUESTS sent has been scheduled. */
+#ifndef IOT_HTTPS_DEMO_ASYNC_TIMEOUT_MS
+    #define IOT_HTTPS_DEMO_ASYNC_TIMEOUT_MS    ( ( uint32_t ) 300000 )   /* 5 minute timeout for this demo. */
```

Comment:
This would be a pretty long time for a customer to want to wait for one of our demos to complete. Have you encountered this kind of delay in testing?

Reply:
No I have not encountered that time limit. I just wanted to make it this long just in case the server take a long time to respond. The demo WILL finish before 5 minutes.